Different behaviour regarding JUnit 5 @Nested classes between Gradle 4.6 and 4.7-4.8

When trying to run a single test from the `@Nested` class in JUnit 5, the outcome is different between Gradle versions, with 4.6 finding and executing the test successfully, and 4.7 and 4.8 being unable to find the test at all.

### Expected Behavior
The specified test methods in the nested class should be executed.

### Current Behavior
The specified methods are not being found.

### Context
The issue originally arose while I was trying to migrate some tests from JUnit 4 to JUnit 5, replacing the `@RunWith(Enclosed.class)` approach to nested tests we've used before with the new `@Nested` classes in JUnit 5. When I tried to launch these new tests from Intellij IDEA (using those little green triangles on the side), I got a failure like this one (please check the `Steps to Reproduce` section for the test setup explanation)

    Executing tasks ':cleanTest :test --tests "EnclosingClass$NestedClass.nestedTest"'...
    > Task :cleanTest
    > Task :compileJava UP-TO-DATE
    > Task :processResources NO-SOURCE
    > Task :classes UP-TO-DATE
    > Task :compileTestJava UP-TO-DATE
    > Task :processTestResources NO-SOURCE
    > Task :testClasses UP-TO-DATE
    > Task :test FAILED
    FAILURE: Build failed with an exception.
    * What went wrong:
    Execution failed for task ':test'.
    > No tests found for given includes: [EnclosingClass$NestedClass.nestedTest] filter.includeTestsMatching)
    * Try:
    Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
    * Get more help at https://help.gradle.org
    BUILD FAILED in 0s
    4 actionable tasks: 2 executed, 2 up-to-date
    No tests found for given includes: [EnclosingClass$NestedClass.nestedTest] (filter.includeTestsMatching)
    Tasks execution finished ':cleanTest :test --tests "com.example.EnclosingClass$NestedClass.nestedTest"'.

 Note that I used the Gradle Test Runner, which basically executes a Gradle command. When using the native IDEA JUnit execution support, the test is executed normally. This fact made me curious, so I've  tried to launch the tests directly from the command line:

    gradlew cleanTest test --tests EnclosingClass$NestedClass.nestedTest

As expected, the execution failed in a similar manner - though the error message was slightly different: 
`No tests found for given includes: [EnclosingClass$NestedClass.nestedTest] (--tests filter)`

And here's the interesting part:
- when using Gradle 4.6 instead of 4.7 (or 4.8 for that matter) the nested test is executed properly
```
gradlew :cleanTest :test --tests "EnclosingClass$NestedClass.nestedTest"

> Task :test
EnclosingClass > EnclosingClass$NestedClass.nestedTest() STARTED
EnclosingClass > EnclosingClass$NestedClass.nestedTest() PASSED
BUILD SUCCESSFUL in 1s
4 actionable tasks: 4 executed
```
- If a wildcard is added in front of the filter, then the test gets successfully found and executed too:
`gradlew cleanTest test --tests "*com.example.EnclosingClass$NestedClass.nestedTest"`

I suspect that this issue can be related to the changes in https://github.com/gradle/gradle/pull/4597, though it's mostly because of the leading wildcard changing the behaviour.

### Steps to Reproduce (for bugs)
Here's a very simple setup the demonstrates this behaviour:
- First, do a `gradle init --type java-library`
- Replace the contents of build.gradle with the following:
```
plugins {
    id 'java'
}

dependencies {
    testCompile 'org.junit.jupiter:junit-jupiter-api:5.2.0'
    testCompile 'org.junit.jupiter:junit-jupiter-params:5.2.0'
    testRuntime 'org.junit.jupiter:junit-jupiter-engine:5.2.0'
}

repositories {
    jcenter()
}

test {
    useJUnitPlatform()
    testLogging {
        events "STARTED", "PASSED", "FAILED", "SKIPPED"
    }
}
```
- Create a new source file `src/java/test/EnclosingClass.java` with the following contents:
```
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Nested;
import static org.junit.jupiter.api.Assertions.assertEquals;

class EnclosingClass {
    @Nested
    class NestedClass {
        @Test
        void nestedTest() {
            assertEquals(2, 1 + 1);
        }
    }
}
```
- Make sure that we're using Gradle 4.6 in `gradle/wrapper/gradle-wrapper.properties`
- Execute the following command line and note that the test is found and executed:
 `gradlew cleanTest test --tests EnclosingClass$NestedClass.nestedTest`
- Change Gradle version to 4.7 and execute the same command. Note that the test is not found.

### Your Environment
Build scan, Gradle 4.7, test detection failed: https://scans.gradle.com/s/i3licwrkey3pa
The same for 4.6, test detection successful: https://gradle.com/s/nzbs6as64cb6a

--- a/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java
+++ b/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java
@@ -158,7 +158,7 @@ public class TestSelectionMatcher {
         }
 
         private boolean lastClassNameElementMatchesPenultimatePatternElement(String[] className, int index) {
-            return index == segments.length - 2 && index == className.length - 1 && className[index].equals(segments[index]);
+            return index == segments.length - 2 && index == className.length - 1 && classNameMatch(className[index], segments[index]);
         }
 
         private boolean lastClassNameElementMatchesLastPatternElement(String[] className, int index) {
@@ -186,6 +186,18 @@ public class TestSelectionMatcher {
         return simpleName;
     }
 
+    // Foo can match both Foo and Foo$NestedClass
+    // https://github.com/gradle/gradle/issues/5763
+    private static boolean classNameMatch(String simpleClassName, String patternSimpleClassName) {
+        if (simpleClassName.equals(patternSimpleClassName)) {
+            return true;
+        } else if (patternSimpleClassName.contains("$")) {
+            return simpleClassName.equals(patternSimpleClassName.substring(0, patternSimpleClassName.indexOf('$')));
+        } else {
+            return false;
+        }
+    }
+
     private interface LastElementMatcher {
         boolean match(String classElement, String patternElement);
     }
@@ -193,14 +205,14 @@ public class TestSelectionMatcher {
     private static class NoWildcardMatcher implements LastElementMatcher {
         @Override
         public boolean match(String classElement, String patternElement) {
-            return classElement.equals(patternElement);
+            return classNameMatch(classElement, patternElement);
         }
     }
 
     private static class WildcardMatcher implements LastElementMatcher {
         @Override
         public boolean match(String classElement, String patternElement) {
-            return classElement.startsWith(patternElement);
+            return classElement.startsWith(patternElement) || classNameMatch(classElement, patternElement);
         }
     }
 

--- a/subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcherTest.groovy
+++ b/subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcherTest.groovy
@@ -197,6 +197,12 @@ class TestSelectionMatcherTest extends Specification {
         ['Foo']                           | 'FooTest'            | false
         ['org.gradle.Foo']                | 'org.gradle.FooTest' | false
         ['org.gradle.Foo.*']              | 'org.gradle.FooTest' | false
+
+        ['org.gradle.Foo$Bar.*test']      | 'Foo'                | false
+        ['org.gradle.Foo$Bar.*test']      | 'org.Foo'            | false
+        ['org.gradle.Foo$Bar.*test']      | 'org.gradle.Foo'     | true
+        ['Enclosing$Nested.test']         | "Enclosing"          | true
+        ['org.gradle.Foo$1$2.test']       | "org.gradle.Foo"     | true
     }
 
     @Unroll

--- a/subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformNestedPatternMatchingIntegrationTest.groovy
+++ b/subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformNestedPatternMatchingIntegrationTest.groovy
@@ -1,0 +1,60 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.testing.junitplatform
+
+import org.gradle.integtests.fixtures.DefaultTestExecutionResult
+
+class JUnitPlatformNestedPatternMatchingIntegrationTest extends JUnitPlatformIntegrationSpec {
+    def 'can use nested class as test pattern'() {
+        given:
+        file('src/test/java/EnclosingClass.java') << '''
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Nested;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class EnclosingClass {
+    @Nested
+    class NestedClass {
+        @Test
+        void nestedTest() {
+        }
+        @Test
+        void anotherTest() {
+        }
+    }
+    @Nested
+    class AnotherNestedClass {
+        @Test
+        void foo() {
+        }
+    }
+    @Test
+    void foo() {
+    }
+}
+'''
+        when:
+        succeeds('test', '--tests', 'EnclosingClass$NestedClass.nestedTest')
+
+        then:
+        new DefaultTestExecutionResult(testDirectory)
+            .assertTestClassesExecuted('EnclosingClass$NestedClass')
+            .testClass('EnclosingClass$NestedClass')
+            .assertTestCount(1, 0, 0)
+            .assertTestPassed('nestedTest')
+    }
+}

